### DISCUSSION TITLE
Using Libigl mesh_boolean to fix meshes

### DISCUSSION THREAD

**Discussion by gsohler** [libigl/libigl#2180] [Q&A]
lib libigl community.

I have written a function, which can offsed 3D objects with a certain value.
Afterwards i want to use libigl to fix the resulting mesh. (i found out that mesh_boolean RESOLVE is great at fixing self intersections)
With a positive offset value, holes in the object might be closed and i want libigl to remove the extra vertices
With a negative offset value,  small details might disappear and i want libigl to remove the extra vertices.

you can see some examples:

offset 0:  this is the basis
 https://imgpile.com/i/hufXng

positive offset: 
 https://imgpile.com/i/hufHxc
you see with an positive offset the relation in dimension between the cylinder and cube gets smaller

negative offset: 
https://imgpile.com/i/hufoqN
also works fine, you see the the realtion between cylinder and cube gets begger

more negative offset:
 https://imgpile.com/i/hufoqN
But this case has a problem.
The cube is still ok, but the  offset algorithm has shrunk the cylinder beyone non-existance.
the faces  of the cylinder are reversed and actually represent negative volume e.g a cylindrical hole in space.

still libigl displays a positive object instead, which is not my expectation.

I tried to improve libigl with this code in BinaryWindingNumberOperatations.h
but it did not do the trick 

      template <>
      class BinaryWindingNumberOperations<MESH_BOOLEAN_TYPE_ADDUNION> {
        public:
          template<typename DerivedW>
          typename DerivedW::Scalar operator()(
              const Eigen::PlainObjectBase<DerivedW>& win_nums) const 
          {
	    int cnt=0;
            for(int i = 0;i<win_nums.size();i++)
            {
              if(win_nums(i) > 0) cnt++; else cnt--;
            }
            return cnt>0?true:false;
      };

Can anybody tell me, how i can use libigl to remove negative space from an object ?

Thank you for reading

**Comment by alecjacobson** [comment]
Could you provide a minimal working example including how you do the offset? It'd be great to debug on a full example.

**Reply by gsohler** [reply]
Hi Alec,
Thank you for your message. 
I have thought a little bit about how to best provide you with a  genuine testcase and  i have to the conclusion  that it might be the best
to debug libigl "in place".
Would it be ok to you if you compile my github branch, where libigl is (somehow) integrated with my offset in place plus sample
script ?
you might know OpenSCAD. do you ?

**Comment by gsohler** [comment]
Hi
Alternatively i could create a c File testcase with f and v Input matrices.
What die you think?

Alec Jacobson ***@***.***> schrieb am Fr., 10. März 2023,
18:45:

**Reply by gsohler** [reply]
Hi, i was able to create a small testcase for you. 
in the output i expect only to see the cube,  because the prisma has negative space.
but i see the cube and the prisma instead.

sorry for pasting the code here:

===================
#include <stdio.h>
#include <Eigen/Core>
#include <Eigen/Geometry>

#include <igl/copyleft/cgal/mesh_boolean.h>
#include <igl/writeSTL.h>

int main(void)
{
	Eigen::MatrixXd  VA, VB, VC;
	Eigen::MatrixXi FA, FB, FC;
	Eigen::VectorXi J;
	VA.resize(18,3);
	FA.resize(28,3);

	VA <<	1.5, -1.5, 1.5, 
		1.5, 1.5, -1.5, 
		1.5, 1.5, 1.5, 
		1.5, -1.5, -1.5, 
		-0.73607, -0, 1.5, 
		-0.227456, -0.700043, 1.5, 
		-1.5, 1.5, 1.5, 
		0.595492, -0.432649, 1.5, 
		0.595492, 0.432649, 1.5, 
		-0.227456, 0.700043, 1.5, 
		-1.5, -1.5, 1.5, 
		-1.5, -1.5, -1.5, 
		-1.5, 1.5, -1.5, 
		-0.227456, 0.700043, 4, 
		0.595492, 0.432649, 4, 
		0.595492, -0.432649, 4, 
		-0.227456, -0.700043, 4, 
		-0.73607, -0, 4;

	FA <<	1, 2, 3, 
		0, 3, 2, 
		8, 7, 9, 
		7, 5, 9, 
		4, 9, 5, 
		15, 14, 16, 
		14, 13, 16, 
		17, 16, 13, 
		0, 2, 10, 
		6, 10, 2, 
		1, 3, 12, 
		11, 12, 3, 
		6, 12, 10, 
		11, 10, 12, 
		8, 14, 7, 
		15, 7, 14, 
		6, 2, 12, 
		1, 12, 2, 
		0, 10, 3, 
		11, 3, 10, 
		13, 14, 9, 
		8, 9, 14, 
		15, 16, 7, 
		5, 7, 16, 
		17, 4, 16, 
		5, 16, 4, 
		13, 9, 17, 
		4, 17, 9
	;

	
	igl::copyleft::cgal::mesh_boolean(VA,FA,VB,FB,igl::MESH_BOOLEAN_TYPE_RESOLVE,VC,FC,J);

	igl::writeSTL("result.stl",VC,FC);
}
